librz/analysis/arch/x86/x86_il.c

```diff
@@ -0,0 +1,3556 @@
+	stack_mem.segment = X86_REG_SS;
+
+	unsigned int dflag = user_op_size;
+	unsigned int op_size;
+	unsigned int stack_size = bitness / BITS_PER_BYTE;
+
+	if (ins) {
+		if (bitness == 64) {
+			dflag = ins->structure->rex ? 8 : ins->structure->prefix[2] ? 2
+										    : 4; /* in bytes */
+			stack_size = 8; /* in bytes */
+		} else {
+			/* We use the other operand and address size if the prefix is set */
+			if ((bitness == 32) ^ ins->structure->prefix[2]) {
+				dflag = 4;
+			} else {
+				dflag = 2;
+			}
+
+			stack_size = 4;
+		}
+	}
+
+	if (bitness == 64) {
+		op_size = (dflag == 2) ? 2 : 8;
+	} else {
+		op_size = dflag;
+	}
+
+	RzILOpEffect *ret = STOREW(x86_il_get_memaddr_bits(stack_mem, bitness), UNSIGNED(op_size * BITS_PER_BYTE, val));
+	ret = SEQ2(x86_il_set_reg_bits(X86_REG_RSP, SUB(x86_il_get_reg_bits(X86_REG_RSP, bitness), UN(bitness, stack_size)), bitness), ret);
+
+	return ret;
+}
+
+#define x86_push_helper(val, op_size) x86_push_helper_impl(val, op_size, analysis->bits, NULL)
+
+/**
+ * CALL
+ * Perform a function call
+ * Encoding: D, M
+ */
+IL_LIFTER(call) {
+	/*
+	 * The following implementation is not accurate, since there are many nitty-gritties involved.
+	 * Like whether the call is a near or far call, absolute or relative call, etc.
+	 * Implementing it accurately will require exceptions, task switching,
+	 * shadow stack, CPU internal flags, segmentation support
+	 * Just pushing the current program counter is a good approcimation for now
+	 * shadow stack, CPU internal flags, segmentation support
+	 * Just pushing the current program counter is a good approcimation for now
+	 * We also need to push the code segment register in case 16 and 32 bit modes.
+	 */
+
+	if (analysis->bits == 64) {
+		return x86_push_helper(U64(pc), 2);
+	} else {
+		return SEQ4(SETL("_cs", UNSIGNED(analysis->bits, x86_il_get_reg(X86_REG_CS))), x86_push_helper(VARL("_cs"), analysis->bits / BITS_PER_BYTE), SETL("_pc", UN(analysis->bits, pc)), x86_push_helper(VARL("_pc"), analysis->bits / BITS_PER_BYTE));
+	}
+}
+
+/**
+ * PUSH
+ * Push value on the stack
+ * Encoding:
+ *  - M
+ *  - O
+ *  - I
+ *  - ZO
+ */
+IL_LIFTER(push) {
+	return x86_push_helper_impl(x86_il_get_op(0), ins->structure->operands->size, analysis->bits, ins);
+}
+
+/**
+ * PUSHF
+ * Push FLAGS register onto the stack (16 bits)
+ * Encoding: ZO
+ */
+IL_LIFTER(pushf) {
+	return x86_push_helper(x86_il_get_flags(16), 2);
+}
+
+/**
+ * PUSHFD
+ * Push EFLAGS register onto the stack (32 bits)
+ * Encoding: ZO
+ */
+IL_LIFTER(pushfd) {
+	return x86_push_helper(x86_il_get_flags(32), 4);
+}
+
+/**
+ * PUSHFQ
+ * Push RFLAGS register onto the stack (64 bits)
+ * Encoding: ZO
+ */
+IL_LIFTER(pushfq) {
+	return x86_push_helper(x86_il_get_flags(64), 8);
+}
+
+/**
+ * PUSHA
+ * Push all general-purpose registers (16-bits)
+ * Encoding: ZO
+ */
+IL_LIFTER(pushaw) {
+	if (analysis->bits != 16) {
+		return NULL;
+	}
+
+	RzILOpEffect *temp = SETL("_sp", x86_il_get_reg(X86_REG_SP));
```

Comment:
same

Reply:
Both the assembler and disassembler are broken for `PUSHAW`, so there is no way to test it.

Reply:
Done